core/http-auth-aws/src/test/java/software/amazon/awssdk/http/auth/aws/internal/signer/DefaultAwsV4HttpSignerTest.java

```diff
@@ -343,4 +344,337 @@ public void sign_WithPayloadSigningNullAndHttp_FallsBackToPayloadSigning() {
+    }
+
+    @Test
+    public void signAsync_WithEventStreamContentType_DelegatesToEventStreamPayloadSigner() {
+        AsyncSignRequest<? extends AwsCredentialsIdentity> request = generateBasicAsyncRequest(
+            AwsCredentialsIdentity.create("access", "secret"),
+            httpRequest -> httpRequest
+                .putHeader("Content-Type", "application/vnd.amazon.eventstream"),
+            signRequest -> {
+            }
+        );
+
+        AsyncSignedRequest signedRequest = signer.signAsync(request).join();
+
+        assertThat(signedRequest.payload().get()).isInstanceOf(SigV4DataFramePublisher.class);
+    }
+
+    @Test
+    public void signAsync_WithEventStreamContentTypeAndUnsignedPayload_Throws() {
+        AsyncSignRequest<? extends AwsCredentialsIdentity> request = generateBasicAsyncRequest(
+            AwsCredentialsIdentity.create("access", "secret"),
+            httpRequest -> httpRequest
+                .putHeader("Content-Type", "application/vnd.amazon.eventstream"),
+            signRequest -> signRequest
+                .putProperty(PAYLOAD_SIGNING_ENABLED, false)
+        );
+
+        assertThrows(UnsupportedOperationException.class, () -> signer.signAsync(request));
+    }
+
+    @Test
+    public void signAsync_WithPayloadSigningFalseAndHttp_FallsBackToPayloadSigning() {
+        AsyncSignRequest<? extends AwsCredentialsIdentity> request = generateBasicAsyncRequest(
+            AwsCredentialsIdentity.create("access", "secret"),
+            httpRequest -> httpRequest.uri(URI.create("http://demo.us-east-1.amazonaws.com")),
+            signRequest -> signRequest
+                .putProperty(PAYLOAD_SIGNING_ENABLED, false)
+        );
+
+        AsyncSignedRequest signedRequest = signer.signAsync(request).join();
+
+        assertThat(signedRequest.request().firstMatchingHeader("x-amz-content-sha256"))
+            .hasValue("a15c8292b1d12abbbbe4148605f7872fbdf645618fee5ab0e8072a7b34f155e2");
+    }
+
+    @Test
+    public void signAsync_WithPayloadSigningNullAndHttp_FallsBackToPayloadSigning() {
+        AsyncSignRequest<? extends AwsCredentialsIdentity> request = generateBasicAsyncRequest(
+            AwsCredentialsIdentity.create("access", "secret"),
+            httpRequest -> httpRequest.uri(URI.create("http://demo.us-east-1.amazonaws.com")),
+            signRequest -> signRequest
+                .putProperty(PAYLOAD_SIGNING_ENABLED, null)
+        );
+
+        AsyncSignedRequest signedRequest = signer.signAsync(request).join();
+
+        assertThat(signedRequest.request().firstMatchingHeader("x-amz-content-sha256"))
+            .hasValue("a15c8292b1d12abbbbe4148605f7872fbdf645618fee5ab0e8072a7b34f155e2");
+    }
+
+    // TODO(sra-identity-and-auth): Once chunk-encoding is implemented in the async path, the assertions this test makes should
+    //  be different - it should asserted that signed-payload (STREAMING-AWS4-HMAC-SHA256-PAYLOAD) is the fallback.
```

Comment:
Thinking about the 2 customers of this library:
* AWS SDKs - since this is existing behavior for AWS SDK, we can figure out what it means for SDKs if/when we change this.
* Direct http-auth-aws customers. We'll get these customers when we release these modules. I'm afraid of them also getting this undesirable behavior and increasing the risk/surface area when we make that change later. Would we just consider this a bugfix or is that reason to consider not making this SdkPublicApi yet or some other guidance to let customer know about issues/limiations.

cc @millems @cenedhryn

Reply:
This is behavior that is in the SDK today, the test just asserts that the signing happens in the same way instead of throwing an exception (for not implemented or unsupported).